axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,208 @@ def strategy(self, opponent: Player) -> Action:
+        two_turns_after_good_defection_ratio and the probability values, 
+        as well as sets the value of num_turns_after_good_defection.
+        """
+        self.current_score = compute_final_score(zip(self.history, opponent.history))
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+                ((self.two_turns_after_good_defection_ratio 
+                * self.two_turns_after_good_defection_ratio_count) 
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - ((self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]]))) 
+                / (self.two_turns_after_good_defection_ratio_count + 1)
+                )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+                ((self.one_turn_after_good_defection_ratio 
+                * self.one_turn_after_good_defection_ratio_count)
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - (self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]])) 
+                / (self.one_turn_after_good_defection_ratio_count + 1)
+                )
+            self.one_turn_after_good_defection_ratio_count += 1
+        else:
+            if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+                self.probability = (
+                    (.95 - (((self.one_turn_after_good_defection_ratio)
+                    + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                    + (1 / (((len(self.history))+1) ** 2))
+                    - (self.dict[opponent.history[-1]] / 4)
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "good"
+            elif (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+                self.probability = (
+                    (.25 + (opponent.cooperations / ((len(self.history)) + 1)))
+                    - (self.opponent_consecutive_defections * .25) 
+                    + ((self.current_score[0] 
+                    - self.current_score[1]) / 100) 
+                    + (4 / ((len(self.history)) + 1))
+                    )
+                self.probability = round(self.probability, 4)
+                self.score = "average"
+
+    def strategy(self, opponent: Player) -> Action:
+
+        current_score = compute_final_score(zip(self.history, opponent.history))
+
+        if len(self.history) == 0:
+            return C
+        else: 
+            Tranquilizer.update_state(self, opponent)
+        if opponent.history[-1] == D: 
```

Comment:
This is causing a problem with your current code. Your `update_state` method is also using (for the case when `FD` is 0) `opponent_consecutive_defections` however because of the order in which you have done things this is using the count of consecutive defections from the previous round.

The way I suggest you fix this (and this will require a bit of work) is to leave this where it is but:

- Rename `update_state` to be `update_cooperative_state` and ONLY do the cases corresponding to `FD==1` and `FD==2` there.
- Move the code for the other state (the probabilistic one for the calculation of `self.probability`) to a method called `update_stochastic_state` and that can be called after you have updated `self.opponent_consecutive_defections`.